### ISSUE TITLE
Cross-referencing a function that returns a pointer with :cpp:func: causes error "Invalid definition: Expected end of definition"

### ISSUE THREAD

**Issue by daniellovell** [sphinx-doc/sphinx#6208]
**Describe the bug**
While using the :cpp:func: directive to cross-reference a C++ function that returns a pointer, you will receivean error that the * character in the reference is "unparseable". This makes it impossible to cross-reference a function with a pointer return value

**To Reproduce**

Source (C++): 
```
class Foo
{
    public:
        Foo* Bar();
}   
```
Sphinx (reStructuredText):
```
:cpp:func:`Foo* Foo::Bar()`
```

Result:
```
WARNING: Unpareseable C++ cross-reference: 'Foo*'
Invalid definition: Expected end of definition. [error at 4]
   Foo*
   ---^
```

**Expected behavior**
The expected behavior is to output a properly cross-referenced link to the Foo* Foo::Bar() function.

**Your project**
N/A

**Screenshots**
N/A

**Environment info**
- OS: [e.g. Unix/Linux/Mac/Win/other with version]
- Python version: 3.7.0
- Sphinx version: 2.0.0+
- Sphinx extensions:  breathe
- Extra tools: N/A

**Comment by jakobandersen** [comment]
Thanks for reporting, this is indeed a bug. Note though that there are two formats for function references: (1) a full signature (as you write), for referencing a specific overload, and (2) just the name of the function (here ``Foo::Bar``), which references an arbitrary overload. Format (2) should still work.

**Comment by daniellovell** [comment]
Hi Jakob, thanks for taking a look at the issue. I can confirm that the format (2) you described works as intended, and even most full signatures work correctly with the exception of functions that return pointers.

**Comment by jakobandersen** [comment]
This should be fixed in version 2.0.1.
